File: tests/python/test_prototype_model_acceptance.py
import json
from pathlib import Path

import jsonschema
import yaml


ROOT = Path(__file__).resolve().parents[2]


def test_i24_contract_is_zero_compute_and_lineage_locked():
    config = yaml.safe_load((ROOT / "config/prototype_model_acceptance.yml").read_text())
    assert config["scientific"]["gate_mode"] == "read_only_zero_compute"
    assert config["identity"]["model_validation_id"] == "pmv_1d5412a7b035635a4187fbf6"
    assert config["identity"]["best_checkpoint_sha256"] == "a17477a647d68024cb59ce6c3ce66a703e12143f37340b90c82cd3549b303704"
    assert config["scientific"]["retrieval"]["original_relevance_metrics"] == "forbidden"


def test_i24_schema_rejects_nonzero_compute(tmp_path):
    schema = json.loads((ROOT / "config/schemas/prototype_model_acceptance.schema.json").read_text())
    instance = {
        "schema_version": "1.0.0", "status": "PASS", "model_acceptance_id": "pma_" + "0" * 24,
        "direct_parents": [], "forwarded_lineage": [],
        "checkpoint": {"name": "epoch-005.pt", "path": "/x", "size_bytes": 1,
                       "sha256": "a17477a647d68024cb59ce6c3ce66a703e12143f37340b90c82cd3549b303704",
                       "epoch": 5, "optimizer_step": 40, "read_only": True},
        "gates": {}, "contracts": {}, "zero_compute": {
            "status": "PASS", "additional_optimizer_steps": 1, "forward_calls": 0,
            "augmentation_calls": 0, "state_update_calls": 0,
            "checkpoint_sha256_before": "0" * 64, "checkpoint_sha256_after": "0" * 64},
        "scientific_identity": {}, "immutable_publication": {
            "atomic": "PASS", "identical_rebuild_reuse": "PASS",
            "same_id_different_content_hard_failure": "PASS"}, "outputs": []}
    try:
        jsonschema.validate(instance, schema)
    except jsonschema.ValidationError:
        return
    raise AssertionError("schema accepted a nonzero optimizer-step count")

File: tests/testthat/test-prototype-model-acceptance.R
test_that("I24 is a read-only gate with the blueprint direct parents", {
  config <- yaml::read_yaml(file.path(fuse_test_root, "config/prototype_model_acceptance.yml"))
  expect_identical(config$scientific$gate_mode, "read_only_zero_compute")
  expect_setequal(
    unlist(config$scientific$direct_parents, use.names = FALSE),
    c("prototype_dataloader_smoke", "prototype_encoder_smoke", "prototype_augmentation_benchmark",
      "prototype_training_acceptance", "prototype_model_validation")
  )
  expect_true(all(c("forward", "augmentation", "optimizer_update", "scheduler_update",
                    "ema_update", "queue_update", "checkpoint_mutation") %in%
                  unlist(config$scientific$forbidden_operations, use.names = FALSE)))
  expect_identical(config$scientific$retrieval$original_relevance_metrics, "forbidden")
  expect_identical(config$scientific$retrieval$augmented_source_metrics,
                   c("MRR", "HIT@1", "HIT@5", "HIT@10"))

  manifest <- targets::tar_manifest(script = file.path(fuse_test_root, "_targets.R"),
                                    callr_arguments = list(wd = fuse_test_root))
  network <- targets::tar_network(
    script = file.path(fuse_test_root, "_targets.R"), callr_arguments = list(wd = fuse_test_root)
  )
  parents <- network$edges$from[network$edges$to == "prototype_model_acceptance"]
  parents <- intersect(parents, manifest$name)
  expect_setequal(parents, c("prototype_dataloader_smoke", "prototype_encoder_smoke",
                             "prototype_augmentation_benchmark", "prototype_training_acceptance",
                             "prototype_model_validation", "prototype_model_acceptance_contract_files"))
  expect_false(any(grepl("^full_|^single_gpu_experiment|^ddp_experiment", manifest$name)))
})

test_that("I24 schema fixes zero-compute and immutable publication gates", {
  schema <- jsonlite::read_json(
    file.path(fuse_test_root, "config/schemas/prototype_model_acceptance.schema.json"),
    simplifyVector = FALSE
  )
  expect_identical(schema$properties$zero_compute$properties$additional_optimizer_steps$const, 0L)
  expect_identical(schema$properties$zero_compute$properties$forward_calls$const, 0L)
  expect_identical(schema$properties$zero_compute$properties$augmentation_calls$const, 0L)
  expect_identical(schema$properties$checkpoint$properties$sha256$const,
                   "a17477a647d68024cb59ce6c3ce66a703e12143f37340b90c82cd3549b303704")
  expect_identical(schema$properties$immutable_publication$properties$atomic$const, "PASS")
})

test_that("I24 publisher cannot invoke model or training computation", {
  source <- readLines(file.path(fuse_test_root, "python/accept_prototype_model.py"), warn = FALSE)
  text <- paste(source, collapse = "\n")
  expect_false(grepl("import torch", text, fixed = TRUE))
  expect_false(grepl("prototype_encoder", text, fixed = TRUE))
  expect_false(grepl("prototype_augmentation", text, fixed = TRUE))
  expect_false(grepl("optimizer.step", text, fixed = TRUE))
})
